### DISCUSSION TITLE
Checking if not-nullable reference is null should produce a warning

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#4216] [General]
Consider the following:
```cs
byte[] foo() => new byte[10];
......
byte[] a = new byte[10];
if (a != null) // we know "a" is not null, why would we check for null?
	Console.WriteLine("");
else
	Console.WriteLine(""); // this is unreachable code, from the perspective of nullability analysis

byte[] b = foo();
if (b != null) // same - we know "b" is not null, why would we check for null?
	Console.WriteLine("");
else
	Console.WriteLine(""); // same - this is unreachable code, from the perspective of nullability analysis
```
I think there should be warnings for these situations, either "unrechable code" or "invalid comparison" - similarly to when comparing a value type to null.
Warning	CS0472	The result of the expression is always 'false' since a value of type 'int' is never equal to 'null' of type 'int?'	
Warning	CS0162	Unreachable code detected

Granted these exact warnings shouldn't be used, because at runtime it's possible that b becomes null, however I do think two new warnings are called for.

**Comment by canton7** [comment]
The nullability analysis actually took the opposite route.

```cs
public void M(string s) {
    Console.WriteLine(s.Length); // Fine

    if (s == null)
    {
        Console.WriteLine("Null");
    }

    Console.WriteLine(s.Length); // Warning
}
```

The nullability analysis learns from your `if` statements: if you test something for null, it figures that you probably have good reason to, and updates the null state for that variable.

**Reply by TahirAhmadov** [reply]
I know, and while I personally disagree with this, it's probably too late to undo it. However, we can add the warnings I talked about in addition to existing logic: when you check not-null variable for null, you get a warning; and it's marked as nullable for statements below.

**Reply by 333fred** [reply]
We explicitly did not do this because we consider adding non-nullability warnings to be a breaking change.

**Reply by TahirAhmadov** [reply]
> We explicitly did not do this because we consider adding non-nullability warnings to be a breaking change.

Wasn't the whole nullability change a huge breaking change? Now that a compiler flag is introduced which breaks things, what's one more warning?

**Reply by 333fred** [reply]
> > We explicitly did not do this because we consider adding non-nullability warnings to be a breaking change.
> 
> Wasn't the whole nullability change a huge breaking change? Now that a compiler flag is introduced which breaks things, what's one more warning?

No. Nullability only added new nullability-related warnings, and only if you opt in. We're also very concerned about false positives: there are many good reasons to check if a variable is possibly null even if its current flow state is "not null". You could be guarding a public api, for example, or you know that some external method was misannotated.

**Reply by TahirAhmadov** [reply]
> No. Nullability only added new nullability-related warnings, and only if you opt in. We're also very concerned about false positives: there are many good reasons to check if a variable is possibly null even if its current flow state is "not null". You could be guarding a public api, for example, or you know that some external method was misannotated.

For public API scenario, you'd just `if(a==null) throw new ArgumentNullException(nameof(a));`, so the rest of the method body has this parameter as not-nullable. 
For external method potentially returning null and incorrectly annotated as not-null return, just assign to nullable local, then check for null, etc.: `string? x = Foo(); if(x == null) { ... }`
How about an IDE code style message for this?

**Reply by 333fred** [reply]
Ignoring for a moment that the declared nullability of a variable does not affect its starting flow state. There could be a valid reason for declaring that local as nullable, maybe you reassign it to a nullable value later, even though it starts as not nullable. How would the code style message differentiate these cases?

**Reply by TahirAhmadov** [reply]
> Ignoring for a moment that the declared nullability of a variable does not affect its starting flow state. There could be a valid reason for declaring that local as nullable, maybe you reassign it to a nullable value later, even though it starts as not nullable. How would the code style message differentiate these cases?

I'm confused. Are you talking about the 1st question (parameter) or 2nd (external return value)?
For the 1st, the logic would exclude simple `if(a==null)throw ...;` statements.
For the 2nd, why not use the nullability of the variable declaration as a hint to start it off as a nullable flow state, even if a not-null expression is assigned to it?

**Reply by 333fred** [reply]
> For the 2nd, why not use the nullability of the variable declaration as a hint to start it off as a nullable flow state, even if a not-null expression is assigned to it?

For the exact reason I mentioned: there are valid reasons to start with a not null value, but assign it a null value later. It would be generally confusing if we respected the results of flow analysis only sometimes.

**Reply by TahirAhmadov** [reply]
@333fred OK I ran across this problem again. Sorry but I think this was a mistake.
```cs
void foo(object x)
{
  if(x == null) { /* ...logic... */ } else { /*...logic...*/ } // no warning
}
foo(null); // warning
```
The problem above is misannotated parameter x; the call is correct. But according to current rules, I get the warning where the code is correct, and no warning where the code is incorrect. I think both places should get a warning - for checking a notnull for null, and when passing null for notnull parameter.

For your example, this is how it should work:
```cs
void foo(object x)
{
// options 1: suppress
#pragma warning blabla
  if(x == null) throw new ArgumentNullException(nameof(x)); // option 2: special compiler logic - no warning because we're enforcing declarative nullability
  if(x == null) // warning - checking not-nullable value for null to trigger some logic other than a single throw
  {
    // ... logic ...
  }
}
```

**Reply by 333fred** [reply]
The compiler cannot save you from misannotated code. That exact pattern is extremely common in public API where you annotate as not accepting null but still perform a defensive null check because you don't control the input.

**Reply by TahirAhmadov** [reply]
> 
> 
> > For the 2nd, why not use the nullability of the variable declaration as a hint to start it off as a nullable flow state, even if a not-null expression is assigned to it?
> 
> For the exact reason I mentioned: there are valid reasons to start with a not null value, but assign it a null value later. It would be generally confusing if we respected the results of flow analysis only sometimes.

Just so that we understand what's being discussed here. The below is how I would expect it to work.
```cs
void foo(object x) { /* ... */ }

object? x = new object(); // no problem
foo(x); // flow analysis - x is not null, therefore no problem
x = null; // no problem
foo(x); // flow analysis - x may be null, therefore warning
if(x == null) { /* ... */ } // no problem

object x = new object(); // no problem
foo(x); // no need for flow analysis; x is not null, period; no problem
x = null; // warning: null assignment
foo(x); // no need for flow analysis; x is not null, period; no problem
if(x == null) { /* ... */ } // warning: unreachable code

var x = new object(); // var = object (not nullable); everything works same as above
```

**Reply by TahirAhmadov** [reply]
> 
> 
> The compiler cannot save you from misannotated code. That exact pattern is extremely common in public API where you annotate as not accepting null but still perform a defensive null check because you don't control the input.

I'm not disputing that. All that needs to happen, like I showed in my example, if either 1) suppress the "unreachable code" warning and throw if null; or 2) the compiler knows that a single throw when checking a notnull value for null is "defensive" and should not be treated as unreachable code.

**Reply by 333fred** [reply]
> > 
> > 
> > > For the 2nd, why not use the nullability of the variable declaration as a hint to start it off as a nullable flow state, even if a not-null expression is assigned to it?
> > 
> > For the exact reason I mentioned: there are valid reasons to start with a not null value, but assign it a null value later. It would be generally confusing if we respected the results of flow analysis only sometimes.
> 
> Just so that we understand what's being discussed here. The below is how I would expect it to work.
> ```cs
> void foo(object x) { /* ... */ }
> 
> object? x = new object(); // no problem
> foo(x); // flow analysis - x is not null, therefore no problem
> x = null; // no problem
> foo(x); // flow analysis - x may be null, therefore warning
> if(x == null) { /* ... */ } // no problem
> 
> object x = new object(); // no problem
> foo(x); // no need for flow analysis; x is not null, period; no problem
> x = null; // warning: null assignment
> foo(x); // no need for flow analysis; x is not null, period; no problem
> if(x == null) { /* ... */ } // warning: unreachable code
> 
> var x = new object(); // var = object (not nullable); everything works same as above
> ```

The second pattern you show here is a coding style we intentionally _support_. Non-safety warnings, ie warnings that will not cause a null reference (such as assignment to a variable) are intentionally all CS8600, and we expect that some users, to avoid massive changes to their codebases, will disable that warning and leave these variables unaltered. The feature is about flow analysis, not about ignoring flow analysis.

**Reply by TahirAhmadov** [reply]
> The second pattern you show here is a coding style we intentionally _support_. Non-safety warnings, ie warnings that will not cause a null reference (such as assignment to a variable) are intentionally all CS8600, and we expect that some users, to avoid massive changes to their codebases, will disable that warning and leave these variables unaltered. The feature is about flow analysis, not about ignoring flow analysis.

I think I understand what you mean, you're trying to make it easier for people to use most of the not-null improvement without forcing them to pay a high price for it. However, I think there's a mistake in judgement here. The team which disables one of these warnings simply for being annoying, will most likely not turn the feature on in the first place. 

However, even if we accept that line of reasoning, we can simply "re-enable" the flow analysis in this case:
```cs
void foo(object x) { /* ... */ }

object x = new object(); // no problem
foo(x); // no need for flow analysis; x is not null, period; no problem
if(x == null) { /* ... */ } // warning: unreachable code
x = null; // warning: null assignment; also, we turn flow analysis on - x is now known to be "object?"
foo(x); // warning
if(x == null) { /* ... */ } // no problem
```

But, I'd reiterate that it was a mistake to make the language accommodating to "laziness". People can disable all kinds of warnings; bad code will result from it; the language can't bend over backwards to account for bad practices.

**Reply by 333fred** [reply]
The point of the way we designed nullable was to get as many people to adopt it as possible. This means accommodating various coding styles.

> The team which disables one of these warnings simply for being annoying, will most likely not turn the feature on in the first place.

At least one member of the LDM intentionally works this way with their own existing code, and several members of the LDM argued against even allowing locals to be annotated in the first place. We gain nothing by being more restrictive except alienating customers.

**Reply by TahirAhmadov** [reply]
> 
> 
> The point of the way we designed nullable was to get as many people to adopt it as possible. This means accommodating various coding styles.
> 
> > The team which disables one of these warnings simply for being annoying, will most likely not turn the feature on in the first place.
> 
> At least one member of the LDM intentionally works this way with their own existing code, and several members of the LDM argued against even allowing locals to be annotated in the first place. We gain nothing by being more restrictive except alienating customers.

I understand the gradual adoption goal, but the `#nullable enable` directive solves that. Once you enable a file, then you fix the methods over time, then next file, etc. You can even `#nullable disable` part of the file, right? There's no reason to enable nullability for the project, and then demand language do things differently because of the sudden amount of work introduced.

Every decision has pros and cons. This decision, while having a weak pro - the large amount of work needed to fix locals can effectively be reduced using file-by-file (or even by-section) nullability and this "style" isn't really needed - it also has a strong con: new code and in some cases existing code get strange rules going forward forever. Usually, breaking existing code is bad. In this case, though, we're paying this huge price, with a compiler switch, thousands of warnings, etc. Having paid this huge price, we should strive for perfectly.

**Reply by CyrusNajmabadi** [reply]
> I understand the gradual adoption goal, but the #nullable enable directive solves that.

1. I don't think that's true.  
2. Importantly, nullable is fundamentally a *weak* static check (and will remain that way on the current runtime/langauge for the forseeable future), while runtime checks are the only true *strong* runtime check.   As long as the language is weak here, it is important that the strong way of writing the code not be penalized.

**Comment by huoyaoyuan** [comment]
This should be covered by [CA1508](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1508).

**Reply by TahirAhmadov** [reply]
This is a good point. The only exception (no pun intended) should be to throw an exception like ArgumentNullException.

**Reply by TahirAhmadov** [reply]
On second thought, it's not a good point. For `if(x==null) { /* ... */ }`, this works. But for the equally (if not more) common `if(x!=null) { /* ... */ }`, this doesn't; there is no unreachable code here. But we still want to warn about useless comparison.

**Comment by TahirAhmadov** [comment]
I again ran into a bug because of this: in ASP.NET, `Request.Query["p"].ToString()` returns empty string or value, and never null, and I was checking for `null` (because I mistakenly thought `ToString()` returned `string?`).
Is there an analyzer for this? Otherwise, I'm afraid I'll need to write one myself.

**Reply by RikkiGibson** [reply]
dotnet/roslyn#65435 is perhaps relevant (discusses issuing hidden diagnostics as well as actually doing the optimization), but I think it errs more on the side of "it will physically never be possible to observe a null value here", versus "if this API is being honest, we will never get a null value here".

**Reply by TahirAhmadov** [reply]
It is somewhat relevant, but you are right - given that it's discussing a run-time optimization, it has to be "physical". 
I was hoping to find an "honesty"-based analyzer/quick-fixer.

**Reply by RikkiGibson** [reply]
Such an analyzer shouldn't be that hard to write. You would need to visit every expression that is being null-tested and say, if its flow state is already not null (which we have public API for), then report a diagnostic.

**Reply by TahirAhmadov** [reply]
Yes, not hard, but def a few requirements:
- The obvious scenario - `!= null`
- Handle overloaded `==` operators - those should be excluded probably
- `is null` and `is not null` and `is { }`
- `?.`
- `??` and `??=`
- `!.` and `![]`

Plus for completeness, quick fixes to:
- Remove an unneeded `if` statement
- Remove an unneeded condition from a condition expression, handling `&&` and `||` accordingly
- Remove `??=` assignments
- Remove right side of `??` operators

Etc. etc. Not hard, but not very straightforward :)

**Reply by 333fred** [reply]
I doubt it will be as simple as that just visiting and looking at flow states, as I expect a number of those API calls will return the state _after_, not the state _before_. And the state after will be maybe null.

**Reply by TahirAhmadov** [reply]
@333fred is it possible to find the previous statement and call the nullability result API on that one?

**Reply by RikkiGibson** [reply]
Yeah, but I think looking within the expression that does the null test is sufficient. Just look at the state after the operand expression, e.g. `s` in `s == null`. In a method like this one

```cs
void M(string s)
{
    if (s == null)
    {
        s.ToString();
    }
}
```

we expect a not-null state afterwards, which appears to be reflected in the IDE's display of the null state

![image](https://user-images.githubusercontent.com/5833655/205780722-75d8d592-f23e-44ab-a464-d34124439e7c.png)

**Reply by 333fred** [reply]
@RikkiGibson I don't expect that this is result you'll get for every scenario that Tahir will want to cover.

**Reply by 333fred** [reply]
> @333fred is it possible to find the previous statement and call the nullability result API on that one?

If there is one. There is no guarantee there is.

**Comment by TahirAhmadov** [comment]
For those who are interested:

https://www.nuget.org/packages/TA.Analyzers.InvalidNullChecks/
